internal/controlplane/identity_events.go

```diff
@@ -109,71 +109,66 @@ func HandleEvents(
 		Str("operation", "delete").
 		Str("subject", userId).
 		Logger()
 
-	tx, err := store.BeginTransaction()
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if retErr != nil {
-			if err := store.Rollback(tx); err != nil && !errors.Is(err, sql.ErrTxDone) {
-				l.Debug().Msgf("error rolling back transaction: %v", err)
-			}
-		}
-	}()
-	qtx := store.GetQuerierWithTransaction(tx)
-
-	var userDBID *int32
-
-	usr, err := qtx.GetUserBySubject(ctx, userId)
-	// If the user doesn't exist, we still want to clean up any associated data
-	if err != nil && !errors.Is(err, sql.ErrNoRows) {
-		return fmt.Errorf("error retrieving user %v", err)
-	} else if err == nil {
-		userDBID = &usr.ID
-		l = l.With().Int32("user_id", usr.ID).Logger()
-	}
-
 	// Fetching the projects for user before the deletion was made.
 	// This allows us to clean up the project from the database
 	// if there are no more role assignments for the project.
 	projs, err := authzClient.ProjectsForUser(ctx, userId)
 	if err != nil {
 		return fmt.Errorf("error getting projects for user %v", err)
 	}
+	l.Debug().Int("projects", len(projs)).Msg("projects for user")
+
+	dbUser, err := db.WithTransaction(store, func(qtx db.ExtendQuerier) (db.User, error) {
+		var userDBID *int32
+
+		usr, err := qtx.GetUserBySubject(ctx, userId)
+		// If the user doesn't exist, we still want to clean up any associated data
+		if err != nil && !errors.Is(err, sql.ErrNoRows) {
+			return db.User{}, fmt.Errorf("error retrieving user %v", err)
+		} else if err == nil {
+			userDBID = &usr.ID
+			l = l.With().Int32("user_id", usr.ID).Logger()
+		}
 
-	l.Debug().Msg("deleting user from authorization system")
-	// We delete the user from the authorization system first
-	if err := authzClient.DeleteUser(ctx, userId); err != nil {
-		return fmt.Errorf("error deleting authorization tuple %v", err)
-	}
-
-	// We only delete the user if it still exists in the database
-	if userDBID != nil {
-		l.Debug().Msg("deleting user from database")
-		if err := qtx.DeleteUser(ctx, *userDBID); err != nil {
-			return fmt.Errorf("error deleting user %v", err)
+		// We only delete the user if it still exists in the database
+		if userDBID != nil {
+			l.Debug().Msg("deleting user from database")
+			if err := qtx.DeleteUser(ctx, *userDBID); err != nil {
+				return db.User{}, fmt.Errorf("error deleting user %v", err)
+			}
 		}
-	}
 
-	for _, proj := range projs {
-		l.Debug().Str("project_id", proj.String()).Msg("cleaning up project")
-		if err := projects.CleanUpUnmanagedProjects(ctx, proj, qtx, authzClient, providerService, l); err != nil {
-			return fmt.Errorf("error deleting project %v", err)
+		for _, proj := range projs {
+			l.Debug().Str("project_id", proj.String()).Msg("cleaning up project")
+			if err := projects.CleanUpUnmanagedProjects(ctx, userId, proj, qtx, authzClient, providerService, l); err != nil {
+				return db.User{}, fmt.Errorf("error deleting project %v", err)
+			}
 		}
```

Comment:
Q: Do we want to return an error here after the first Unmanaged project fails?  Would we want to delete the projects before the user?

A: It turns out the order doesn't matter, because it's happening in a transaction.  It would still feel slightly better to me to delete the user after the projects, but the real effect is minimal.